### PR TITLE
Add vector commitment

### PR DESCRIPTION
**Pull request by algoidan**
## Summary
We add a vector commitment implementation based on the existing merkle tree.
this Implementation will bind order to the commitments which are crucial for StateProofs

### CODE REVIEW

File: crypto/merklearray/array.go [hunk 1]
```diff
@@ -16,10 +16,12 @@
 
 package merklearray
 
+import "github.com/algorand/go-algorand/crypto"
+
 // An Array represents a dense array of leaf elements that are
 // combined into a Merkle tree. The Marshal method returns a byte slice that represents the object
 // that the Tree will use to hash the leaves.
```
Comment (algonautshant) [inline crypto/merklearray/array.go:23]: This comment is very vague. Please explain what "represents" means with more details. This is an exported interface, so the code will be much more readable when this interface is well understood.

File: crypto/merklearray/vectorCommitmentArray.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+type bottomElement struct{}
+
+func (b *bottomElement) ToBeHashed() (protocol.HashID, []byte) {
+	return protocol.MerkleVectorCommitmentBottomLeaf, []byte{}
+}
+
+func generateVectorCommitmentArray(innerArray Array) *vectorCommitmentArray {
+	arrayLen := innerArray.Length()
+	if arrayLen == 0 || arrayLen == 1 {
```
[the remark this answers is not in the document]
Comment (algoidan) [reply crypto/merklearray/vectorCommitmentArray.go:46]: I agree. However, I think it is clearer this way. we explicitly distinguish the edge cases.

File: crypto/merklearray/proof.go [hunk 3]
```diff
@@ -27,21 +31,32 @@ type Proof struct {
...
+
+// GetFixedLengthHashableRepresentation serializes the proof into a sequence of bytes.
 // it basically concatenates all the verification path one after another.
 // The function returns a fixed length array for each hash function. which is 1 + MaxTreeDepth * digestsize
 //
 // the path is guaranteed to be less than MaxTreeDepth and if the path length is less
 // than MaxTreeDepth, array is padded with zeros.
 // more details could be found in the Algorand's spec.
-func (p *Proof) GetSerializedProof() []byte {
+func (p *SingleLeafProof) GetFixedLengthHashableRepresentation() []byte {
```
[the remark this answers is not in the document]
Comment (algoidan) [reply crypto/merklearray/proof.go:54]: `GetFixedLengthHashableRepresentation` is a common name for some elements that we need a fixed-length encoding for the SNARK prover.

File: crypto/merklearray/merkle.go [hunk 4]
```diff
@@ -51,11 +51,19 @@ var (
...
-	VectorCommitment bool               `codec:"vc"`
+	// Levels represents the tree in layers. layer[0] contains the leaves.
+	Levels []Layer `codec:"lvls,allocbound=MaxTreeDepth+1"`
+
+	// NumOfLeaves represents the number of the leaves in the tree.
+	// in case the tree represents a vector commitment this value represents
+	// the padded number of leaves (i.e the number of leaves is padded to 2^d)
+	NumOfLeaves uint64 `codec:"nl"`
+
+	// Hash represents the hash function which is being using on elements in this tree.
```
Comment (algonautshant) [inline crypto/merklearray/merkle.go:62]: typo: using -> used

File: crypto/merklearray/merkle.go [hunk 5]
```diff
@@ -33,14 +34,36 @@ const (
...
+	Hash crypto.HashFactory `codec:"hsh"`
+
+	// IsVectorCommitment determines whether the tree was built as a vector commitment
+	IsVectorCommitment bool `codec:"vc"`
```
[the remark this answers is not in the document]
Comment (algoidan) [reply crypto/merklearray/merkle.go:66]: all fields should be public for the msgpack

File: crypto/merklearray/merkle.go [hunk 6]
```diff
@@ -208,41 +300,55 @@ func (tree *Tree) buildNextLayer() {
 	tree.Levels = append(tree.Levels, newLayer)
 }
 
-func hashLeaves(elems map[uint64]crypto.Hashable, hash hash.Hash) (map[uint64]crypto.GenericDigest, error) {
+// VerifyVectorCommitment verifies a vector commitment proof against a given root.
+func VerifyVectorCommitment(root crypto.GenericDigest, elems map[uint64]crypto.Hashable, proof *Proof) error {
```
[the remark this answers is not in the document]
Comment (algonautshant) [reply crypto/merklearray/merkle.go:304]: VerifyVectorCommitment will be used for generic Proof with multiple leaves in the future. Resolving this issue.

File: crypto/merklearray/proof.go [hunk 7]
```diff
@@ -27,21 +31,32 @@ type Proof struct {
 	// the path length can increase up to 2^MaxTreeDepth / 2
 	Path        []crypto.GenericDigest `codec:"pth,allocbound=MaxNumLeaves/2"`
 	HashFactory crypto.HashFactory     `codec:"hsh"`
+	// TreeDepth represents the depth of the tree that is being proven.
+	// It is the number of edges from the root to a leaf.
+	TreeDepth uint8 `codec:"td"`
```
[the remark this answers is not in the document]
Comment (algonautshant) [reply crypto/merklearray/proof.go:36]: It is also relevant for proves with more than one leaf!

File: crypto/merklearray/merkle.go [hunk 8]
```diff
@@ -132,28 +177,51 @@ func (tree *Tree) Root() crypto.GenericDigest {
...
+		}
+	}
+
+	if tree.IsVectorCommitment {
+		VcIdxs, err := tree.convertLeavesIndexes(idxs)
+		if err != nil {
+			return nil, err
+		}
+		idxs = VcIdxs
 	}
```
[the remark this answers is not in the document]
Comment (algonautshant) [reply crypto/merklearray/merkle.go:225]: never mind. `convertLeavesIndexes` is used besides `ProveSingleLeaf`.

File: crypto/merklearray/merkle.go [hunk 9]
```diff
@@ -101,8 +143,10 @@ func Build(array Array, factory crypto.HashFactory) (*Tree, error) {
...
 
 	tree := &Tree{
-		Levels: nil,
-		Hash:   factory,
+		Levels:             nil,
+		NumOfLeaves:        array.Length(),
+		Hash:               factory,
+		IsVectorCommitment: false,
```
[the remark this answers is not in the document]
Comment (algonautshant) [reply crypto/merklearray/merkle.go:151]: never mind. `convertLeavesIndexes` is used besides `ProveSingleLeaf`. 
This is a personal preference at this point, and okay to leave it as is.

File: crypto/merklearray/merkle.go [hunk 10]
```diff
@@ -208,41 +300,55 @@ func (tree *Tree) buildNextLayer() {
...
 // returned by Prove().
 func Verify(root crypto.GenericDigest, elems map[uint64]crypto.Hashable, proof *Proof) error {
-
-	if proof == nil {
-		return fmt.Errorf("proof should not be nil")
+	if err := checkInput(proof); err != nil {
+		return err
 	}
 
 	if len(elems) == 0 {
```
[the remark this answers is not in the document]
Comment (algonautshant) [reply crypto/merklearray/merkle.go:333]: the check in `hashLeaves` is a check for the map key. Cannot be checked here.

File: crypto/merklearray/proof.go [hunk 11]
```diff
@@ -16,36 +16,51 @@
...
-// than MaxTreeDepth, array is padded with zeros.
+// the path is guaranteed to be less than MaxEncodedTreeDepth and if the path length is less
+// than MaxEncodedTreeDepth, array is padded with zeros.
 // more details could be found in the Algorand's spec.
-func (p *Proof) GetSerializedProof() []byte {
+func (p *SingleLeafProof) GetFixedLengthHashableRepresentation() []byte {
 	hash := p.HashFactory.NewHash()
 
-	var binProof = make([]byte, 0, 1+(MaxTreeDepth*hash.Size()))
+	var binProof = make([]byte, 0, 1+(MaxEncodedTreeDepth*hash.Size()))
```
Comment (algonautshant) [inline crypto/merklearray/proof.go:57]: Should there be an error in case the tree size is larger than `MaxEncodedTreeDepth`?

**Review by algonautshant — APPROVED** [review]
Looks Great!
Thanks for all the updates.

**Pull request merged**